components/controls/Tooltip.tsx

```diff
@@ -186,12 +185,56 @@ export class TooltipInner extends React.Component<TooltipProps, State> {
 
       // save width and height (and later set in `render`) to avoid resizing the tooltip element,
       // when it's placed close to the window edge
-      const measurements: Measurements = {
+      measurements = {
         left: window.pageXOffset + left,
         top: window.pageYOffset + top,
         width,
         height
       };
+    }
+    return measurements;
+  };
+
+  tooltipNodeRef = (node: HTMLElement | null) => {
+    this.tooltipNode = node;
+  };
+
+  adjustArrowPosition = (placement: Placement, { leftFix = 0, topFix = 0 }: Fixes) => {
+    switch (placement) {
+      case 'left':
+      case 'right':
+        return { marginTop: -topFix };
+      default:
+        return { marginLeft: -leftFix };
+    }
+  };
+
+  adjustTooltipPosition = (
+    placement: Placement,
+    style: Measurements,
+    { leftFix, topFix }: Fixes
+  ): { placement: Placement; style: Measurements } => {
+    const threshold = 0.5 * this.context.theme.rawSizes.grid;
+    if (placement === 'left' && leftFix && leftFix > threshold) {
+      const { left = style.left } = this.getPosition('right') || {};
+      return { placement: 'right', style: { ...style, left } };
+    } else if (placement === 'right' && leftFix && leftFix < -threshold) {
+      const { left = style.left } = this.getPosition('left') || {};
+      return { placement: 'left', style: { ...style, left } };
+    } else if (placement === 'bottom' && topFix && topFix < -threshold) {
+      const { top = style.top } = this.getPosition('top') || {};
+      return { placement: 'top', style: { ...style, top } };
+    } else if (placement === 'top' && topFix && topFix > threshold) {
+      const { top = style.top } = this.getPosition('bottom') || {};
+      return { placement: 'bottom', style: { ...style, top } };
+    } else {
+      return { placement, style };
+    }
```

Comment:
As discussed, in order to reduce the pain of reading this block, you might want to use an "opposite" lookup table and test threshold against absolute value. Just ideas.

Reply:
Pretty hard to digest indeed

Reply:
This block is pretty hard to digest indeed, it's nice if you have a way to make it more readable!